src/data/list/perm.lean

```diff
@@ -880,6 +880,125 @@ begin
   { refine (IH₁ H).trans (IH₂ ((p₁.pairwise_iff _).1 H)),
     exact λ a b h h₁ h₂, h h₂ h₁ }
 end
+lemma perm.take_inter {α} [decidable_eq α] {xs ys : list α} (n : ℕ)
+  (h : xs ~ ys) (h' : ys.nodup) :
+  xs.take n ~ ys.inter (xs.take n) :=
+begin
+  simp only [list.inter] at *,
+  induction h generalizing n,
+  case list.perm.nil
+  { simp only [not_mem_nil, filter_false, take_nil] },
+  case list.perm.cons
+  { cases n; simp only [mem_cons_iff, true_or, eq_self_iff_true, filter_cons_of_pos, 
+                        perm_cons, take, not_mem_nil, filter_false],
+    cases h' with h₁ h₂,
+    convert h_ih h'_a_1 n using 1,
+    apply filter_congr,
+    introv h, simp only [(h'_a x h).symm, false_or], },
+  case list.perm.swap
+  { rcases h' with ⟨_, _, h₁, h₂⟩,
+    rcases h'_a_2 with ⟨_, _, h₁, h₂⟩,
+    have := h'_a_1 _ (or.inl rfl),
+    cases n; simp only [mem_cons_iff, not_mem_nil, filter_false, take],
+    cases n; simp only [mem_cons_iff, false_or, true_or, filter, *, nat.nat_zero_eq_zero, if_true, 
+                        not_mem_nil, eq_self_iff_true, or_false, if_false, perm_cons, take],
+    { rw filter_eq_nil.2, intros, solve_by_elim [ne.symm], },
+    { convert perm.swap _ _ _, rw @filter_congr _ _ (∈ take n h_l),
+      { clear h'_a_1, induction n generalizing h_l; simp only [not_mem_nil, filter_false, take],
+        cases h_l; simp only [mem_cons_iff, true_or, eq_self_iff_true, filter_cons_of_pos, 
+                              true_and, take, not_mem_nil, filter_false, take_nil],
+        cases h'_a_2_a_2,
+        rwa [@filter_congr _ _ (∈ take n_n h_l_tl), n_ih],
+        { introv h, apply h'_a_2_a_1 _ (or.inr h), },
+        { introv h, simp only [(h'_a_2_a_2_a_1 x h).symm, false_or], }, },
+      { introv h, simp only [(h'_a_2_a_1 x h).symm, (h'_a_1 x (or.inr h)).symm, false_or], } } },
+  case list.perm.trans
+  { transitivity,
+    { apply h_ih_a, rwa h_a_1.nodup_iff },
+    { apply perm.filter _ h_a_1, } },
+end
+
+lemma perm.drop_inter {α} [decidable_eq α] {xs ys : list α} (n : ℕ)
+  (h : xs ~ ys) (h' : ys.nodup) :
+  xs.drop n ~ ys.inter (xs.drop n) :=
+begin
+  by_cases h'' : n ≤ xs.length,
+  { let n' := xs.length - n,
+    have h₀ : n = xs.length - n',
+    { dsimp [n'], rwa nat.sub_sub_self, } ,
+    have h₁ : n' ≤ xs.length,
+    { apply nat.sub_le_self },
+    have h₂ : xs.drop n = (xs.reverse.take n').reverse,
+    { rw [reverse_take _ h₁, h₀, reverse_reverse], },
+    rw [h₂],
+    apply (reverse_perm _).trans,
+    rw inter_reverse,
+    apply perm.take_inter _ _ h',
+    apply (reverse_perm _).trans; assumption, },
+  { have : drop n xs = [],
+    { apply eq_nil_of_length_eq_zero,
+      rw [length_drop, nat.sub_eq_zero_iff_le],
+      apply le_of_not_ge h'' },
+    simp [this, list.inter], }
+end
+
+lemma perm.slice_inter {α} [decidable_eq α] {xs ys : list α} (n m : ℕ)
+  (h : xs ~ ys) (h' : ys.nodup) :
+  list.slice n m xs ~ ys.inter (list.slice n m xs) :=
```

Comment:
Could this proof be reduced to the previous two lemmas, by re-expressing `slice` in terms of `take` and `drop` as suggested above?

Reply:
Yes, you're right, that's a nice improvement. Thanks!